prov/efa: introduce extra request "connid header"

wzamazon opened the pull request:
This PR introduced a new extra request "connid header" and use it to identify the sender's identity.

rwespetal reviewed (commented):
I made it through commit ` prov/efa: introduce function rxr_pkt_init_data_from_tx_entry() ` and will pick up where I left off tomorrow.

shijin-aws commented:
bot:aws:retest

rwespetal commented on prov/efa/src/efa.h:
```diff
@@ -356,6 +355,8 @@ struct efa_reverse_av_key {
 	uint32_t connid;
 };
 
+#define EFA_DGRAM_CONNID (0x0)
```
Didn't realize that `EFA_CONNID_NOTAVAIL` was going to get dropped, anyway the comment in the last commit still applies.

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -202,6 +204,19 @@ In the table, `type` is the packet type ID.
+
+| Bit ID | Value | Name | Description | Used by |
+|-|-|-|-|
+| 15     | 0x8000 | CONNID_HDR | This packet has "connid" in header | extra request "connid header" (section 4.4) |
```
Kind of hinted at this in a previous comment, but why does this start at 0x8000?

wzamazon replied:
Because each packet defines its own set of flags, which starts from the beginning. So I think the universal flag should start from the last bit, and going back wards.

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -496,10 +523,13 @@ Table: 3.1 a list of REQ packet flags
 |  5     | 0x20   | REQ_ATOMIC           | This REQ packet is used by an emulated atomic (write,fetch or compare) communication |
+| 15     | 0x8000 | CONNID_HDR           | This REQ packet has the optional connid header |
+
+Note, the CONNID_HDR flag is an universal flag (table 1.4), and is listed here for completency.
```
completeness

wzamazon replied:
fixed in new revision.

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -734,14 +767,26 @@ Table: 3.6 the binary format a CTS packet
-The field `padding` is 4 bytes padding space to align at 8 bytes boundary.
+Note, when protocol v4 was originally introduced. This field was simply a 4-bytes padding space.
+Later, when we introduce the "connid header" extra feature, we re-purposed this field to to store
+connid. Because "connid header" is an exact request, an endpoint is not obligated to comply.
```
extra request

wzamazon replied:
fixed in new revision

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -1219,7 +1287,12 @@ The binary format of a RECEIPT packet is as the following:
+
+The field `multiuse` has been introduced before when introducing the CTS packet (table 3.6).
+It is a multi-purpose field, which can be used to store `connid` or as a padding
+space, depend on whether the CONNID_HDR universal flag is togged in `flags`. See section 4.4
```
toggled

wzamazon replied:
fixed in new revision

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -1281,6 +1354,44 @@ that is not an feature because the peer does not do anything different. Rather,
+
+The "QP collision" problem arise from fact that the EFA device uses the gid + qpn as the unique
+identifier of a peer. Recall that raw address of EFA endpoint is consisted of 3 parts:
+gid + qpn + connid. EFA device only recognizes gid and qpn. The connid was generated by the endpoint
```
I'd expand the acronyms here for clarity

wzamazon replied:
fixed in new revision.

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -1281,6 +1354,44 @@ that is not an feature because the peer does not do anything different. Rather,
+itself during its initialization.
+
+Because of that, it is possible for an endpoint to receive packets from a destroyed QP, which was used
+by an existed client. As can be seen throughout the document, each packet in the EFA RDM communication
```
"by an existed client" -> "by a previous process that used the same queue pair number."

wzamazon replied:
fixed in new revision.

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -1281,6 +1354,44 @@ that is not an feature because the peer does not do anything different. Rather,
+the bit correspond to 
+Also because this is an extra request, an endpoint cannot assume that the peer support it, thus need to be able
+to handle the case that incoming packets does not have sender connection ID in it. In this case, because the request
+is mainly for accident handling, the application can still function but will be less stable. It is up to the
```
accident handling?

wzamazon replied:
fixed in new revision

rwespetal commented on prov/efa/docs/efa_rdm_protocol_v4.md:
```diff
@@ -1281,6 +1354,44 @@ that is not an feature because the peer does not do anything different. Rather,
+
+A universal flag CONNID_HDR (table 1.4) was designated for CONNID in packet header. An implementation is not required to
+set connid. However, when it does include connid in packet header, it need to toggle on the CONNID_HDR flag in
+the `flags` field of the base header. The exact location of connid is different for each packet type.
```
Nice summary of the issue, thanks for this.

rwespetal commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -0,0 +1,59 @@
+
+	if (base_hdr->type == RXR_CTS_PKT) {
+		return (base_hdr->flags & RXR_PKT_CONNID_HDR)
+			? &(rxr_get_cts_hdr(pkt_entry->pkt)->connid)
```
Minor nit: the ternary operator can be dropped if you check `(base_hdr->flags & RXR_PKT_CONNID_HDR)` at the beginning of the function since the base header is consistent for all packets

wzamazon replied:
Good point. I added a separate commit as the last commit in the PR to refactor this function. When return NULL at the beginning if `base_hdr->flags` does not have RXR_PKT_CONNID_HDR.

rwespetal reviewed (commented):
Couple of questions on backwards compatibility and how we handle dgram qkey/when connid is not in the headers, but overall this looks good to me.

wzamazon commented on prov/efa/src/efa.h:
```diff
@@ -566,6 +566,35 @@ bool rxr_peer_need_raw_addr_hdr(struct rdm_peer *peer)
+ * Connection ID is a 4 bytes random integer identifies an endpoint.
+ * Including connection ID in a packet's header allows peer to
+ * identify sender of the packet. It is necessary because device
+ * only report GID+QPN of a received packet, while multiple senders
```
Thank you! fixed in new revision

wzamazon commented on prov/efa/src/efa.h:
```diff
@@ -566,6 +566,35 @@ bool rxr_peer_need_raw_addr_hdr(struct rdm_peer *peer)
+	/* "connid header" is an extra request defined
+	 * in version 4 (the base version).
+	 * Because it is an extra feature,
+	 * an EP will assume the peer does not need
```
I removed this comment as it is misleading. I think I will add a reference to the document here (at a later commit) since the situation complicated

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -192,13 +219,53 @@ void *rxr_pkt_req_raw_addr(struct rxr_pkt_entry *pkt_entry)
+	base_hdr = rxr_get_base_hdr(pkt_entry->pkt);
+	opt_hdr = (char *)pkt_entry->pkt + rxr_pkt_req_base_hdr_size(pkt_entry);
+
+	if (base_hdr->flags & RXR_REQ_OPT_RAW_ADDR_HDR) {
```
I think it is a good suggestion, but maybe in a separate refactor commit.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -239,6 +317,11 @@ int64_t rxr_pkt_req_cq_data(struct rxr_pkt_entry *pkt_entry)
 
 size_t rxr_pkt_req_max_header_size(int pkt_type)
 {
+	/* max_hdr_size does not include optional connid hdr length because
```
Added in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -174,6 +174,7 @@ void rxr_pkt_init_req_hdr(struct rxr_ep *ep,
 	}
 
 	pkt_entry->addr = tx_entry->addr;
+	assert(opt_hdr - pkt_entry->pkt == rxr_pkt_req_hdr_size(pkt_entry));
```
It was useful when I debug the code, so I left it here. I think it won't do any harm.

wzamazon commented on prov/efa/src/rxr/rdm_proto_v4.h:
```diff
@@ -333,6 +333,11 @@ struct rxr_handshake_hdr {
+	uint32_t connid;
+	uint32_t padding; /* padding to 8 bytes boundary */
+};
+
```
Added in new revision

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -55,5 +55,18 @@ uint32_t *rxr_pkt_connid_ptr(struct rxr_pkt_entry *pkt_entry)
+		char *opt_hdr;
+
+		handshake_hdr = rxr_get_handshake_hdr(pkt_entry->pkt);
+		opt_hdr = pkt_entry->pkt + sizeof(struct rxr_handshake_hdr)
```
Good suggestion. I added a new inline function `rxr_get_handshake_opt_connid_hdr` in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -70,3 +71,86 @@ uint32_t *rxr_pkt_connid_ptr(struct rxr_pkt_entry *pkt_entry)
+	 * pkt_entry->send should be allocated successfully
+	 */
+	pkt_entry->send = ofi_buf_alloc(ep->pkt_sendv_pool);
+	assert(pkt_entry->send);
```
Added a warn in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -70,3 +71,86 @@ uint32_t *rxr_pkt_connid_ptr(struct rxr_pkt_entry *pkt_entry)
+	assert(tx_iov_offset < tx_entry->iov[tx_iov_index].iov_len);
+
+	/*
+	 * Copy can be avoid if:
```
fixed in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -70,3 +71,86 @@ uint32_t *rxr_pkt_connid_ptr(struct rxr_pkt_entry *pkt_entry)
+
+	/*
+	 * Copy can be avoid if:
+	 * 1. user provided memory descriptor, or lower provider does not need memory descriptor
```
fixed in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_cmd.c:
```diff
@@ -81,18 +81,19 @@ ssize_t rxr_pkt_post_data(struct rxr_ep *rxr_ep,
 	 */
 	data_pkt->hdr.seg_offset = tx_entry->bytes_sent;
+	data_pkt->hdr.seg_length = MIN(tx_entry->total_len - tx_entry->bytes_sent,
+				       rxr_ep->mtu_size - sizeof(struct rxr_data_hdr));
```
I forgot we have "ep->max_data_payload_size", used it in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_data.c:
```diff
@@ -36,207 +36,6 @@
-	       "Sending an iov count, %zu with payload size: %lu.\n",
-	       i, payload_size);
-	ret = rxr_pkt_entry_send(ep, pkt_entry, tx_entry->send_flags);
-	if (OFI_UNLIKELY(ret)) {
```
Can you elaborate on this?

I think we only resend when `rxr_pkt_entry_send` return -FI_EAGAIN. The resend was handled by the progress engine `rxr_ep_progress_internal`?

rwespetal replied:
Where this comment is there is some code that resets `bytes_sent` `tx_entry->iov_offset` on failure, so that segment can be resent later. It looks like there is a check at the beginning of `rxr_pkt_post_data` for no TX packets available, and all other errors are fatal.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -70,3 +71,86 @@ uint32_t *rxr_pkt_connid_ptr(struct rxr_pkt_entry *pkt_entry)
+		pkt_entry->send->iov[1].iov_base = (char *)tx_entry->iov[tx_iov_index].iov_base + tx_iov_offset;
+		pkt_entry->send->iov[1].iov_len = data_size;
+		pkt_entry->send->desc[1] = tx_entry->desc[tx_iov_index];
+		pkt_entry->send->iov_count = 2;
```
Yes. That is the intention, and judging from the performance number we got with this PR, this change improved `osu_latency` and did not cause degradation on other changes.

rwespetal replied:
Excellent, thanks.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -154,3 +154,101 @@ void rxr_pkt_init_data_from_tx_entry(struct rxr_ep *ep,
+		return pkt_entry->pkt_size - sizeof(struct rxr_readrsp_hdr);
+
+	if (pkt_type >= RXR_REQ_PKT_BEGIN) {
+		assert(pkt_type == RXR_EAGER_MSGRTM_PKT || pkt_type == RXR_EAGER_TAGRTM_PKT ||
```
agreed

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -154,3 +154,101 @@ void rxr_pkt_init_data_from_tx_entry(struct rxr_ep *ep,
+ * @param[in]		data_offset	the offset of the data in the packet in respect
+ *					of the receiving buffer.
+ * @param[in]		pkt_entry	the packet entry that contains data
+ * @param[in]		data		the pointer poiting th the beginning of data
```
fixed in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_base.c:
```diff
@@ -154,3 +154,101 @@ void rxr_pkt_init_data_from_tx_entry(struct rxr_ep *ep,
+		if (bytes_copied != MIN(data_size, rx_entry->cq_entry.len - data_offset)) {
+			FI_WARN(&rxr_prov, FI_LOG_CQ, "wrong size! bytes_copied: %ld\n",
+				bytes_copied);
+			return -FI_EINVAL;
```
used FI_EIO in new revision.

wzamazon commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -2202,7 +2203,7 @@ int rxr_endpoint(struct fid_domain *domain, struct fi_info *info,
 		rxr_ep->mtu_size = RXR_MTU_MAX_LIMIT;
 
-	rxr_ep->max_data_payload_size = rxr_ep->mtu_size - sizeof(struct rxr_data_hdr);
+	rxr_ep->max_data_payload_size = rxr_ep->mtu_size - sizeof(struct rxr_data_hdr) - sizeof(struct rxr_data_opt_connid_hdr);
```
It is always safe to used a smaller max data payload size. so it will not break backward compatibility

wzamazon commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -1996,6 +1996,30 @@ void rxr_ep_progress_internal(struct rxr_ep *ep)
+		 * DATA packet is different because for DATA packet connid is an optional
+		 * header inserted between the mandatory header and the application data.
+		 * Therefore if peer does not use/understand connid, it will take connid
+		 * as applicaiton data thus cause data corruption.
```
fixed in new revision

wzamazon commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -1704,8 +1705,14 @@ static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
-			pkt_entry->addr = peer ? peer->efa_fiaddr : FI_ADDR_NOTAVAIL;
+			connid = rxr_pkt_connid_ptr(pkt_entry);
+			if (!connid) {
+				FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "No connid in packet header");
```
I though about it. I think I will keep a connid map, which is a map between (ahn + qpn) -> latest connid. If the packet does not have connid, we used the connid in the map. In the meantime, FI_WARN about using latest connid can cause data corruption. Does the solution sounds right to you?

rwespetal replied:
That works for me.

rwespetal commented on prov/efa/src/rxr/rdm_proto_v4.h:
```diff
@@ -138,6 +139,11 @@ struct rxr_base_hdr {
+/* Global flags that can be applied on "rxr_base_hdr.flags" */
+
+/* indicate this packet has the sender connid */
+#define RXR_PKT_CONNID_HDR		BIT_ULL(15)
```
Okay. I guess I don't understand why it starts at bit 15.

wzamazon replied:
because `base_hdr->flags` is a 16 bits integer, os bit 15 is the last bit

rwespetal replied:
Saw the other response, makes sense. Please add a comment for that.

wzamazon replied:
Added in new revision

wzamazon commented:
added two commits to the end of the PR

1. refactor `rxr_pkt_connid_ptr` to improve efficiency.
2. added a connid_map to handle backward compatibility
manually tested the connid_map by running fabtests between two nodes: one using main branch, the other using this PR, and the two nodes can communicate.

rwespetal commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -1707,9 +1707,21 @@ static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
+					     "This means the peer is using an older version of libfabric.\n"
+					     "The communication can continue but it is encouraged to use\n"
+					     "a newer version of libfabric. See section 4.4 of\n"
+					     "https://github.com/ofiwg/libfabric/blob/main/prov/efa/docs/efa_rdm_protocol_v4.md\n"
```
I'd drop the URL to our protocol doc as the average user probably won't want to read that.

wzamazon replied:
Thank you! dropped in new revision

rwespetal reviewed (approved):
Looks good, thank you for the updates.

wzamazon commented:
Did some additional test, and did NOT encounter any issue.
